src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+        /// The limit on the length of form keys.
+        /// </summary>
+        public int KeyLengthLimit { get; set; } = DefaultKeyLengthLimit;
+
+        /// <summary>
+        /// The limit on the length of form values.
+        /// </summary>
+        public int ValueLengthLimit { get; set; } = DefaultValueLengthLimit;
+
+        /// <summary>
+        /// Parses an HTTP form body.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
+        /// <returns>The collection containing the parsed HTTP form body.</returns>
+        public async Task<Dictionary<string, StringValues>> ReadFormAsync(CancellationToken cancellationToken = default)
+        {
+            KeyValueAccumulator accumulator = default;
+            while (true)
+            {
+                ReadResult readResult;
+                if (!_pipeReader.TryRead(out readResult))
+                {
+                    readResult = await _pipeReader.ReadAsync(cancellationToken);
+                }
+
+                var buffer = readResult.Buffer;
+
+                if (!buffer.IsEmpty)
+                {
+                    TryParseFormValues(ref buffer, ref accumulator, readResult.IsCompleted, _encoding, KeyLengthLimit, ValueLengthLimit, ValueCountLimit);
+                }
+
+                if (readResult.IsCompleted)
+                {
+                    break;
+                }
+
+                _pipeReader.AdvanceTo(buffer.Start, buffer.End);
+            }
+
+            return accumulator.GetResults();
+        }
+
+        // Used for testing/perf.
+        internal static void TryParseFormValues(ref ReadOnlySequence<byte> buffer, ref KeyValueAccumulator accumulator, bool isFinalBlock)
+        {
+            TryParseFormValues(ref buffer, ref accumulator, isFinalBlock, Encoding.UTF8);
+        }
+
+        internal static void TryParseFormValues(ref ReadOnlySequence<byte> buffer, ref KeyValueAccumulator accumulator, bool isFinalBlock, Encoding encoding)
+        {
+            TryParseFormValues(ref buffer, ref accumulator, isFinalBlock, encoding, DefaultKeyLengthLimit, DefaultValueLengthLimit, DefaultValueCountLimit);
+        }
+
+        internal static void TryParseFormValues(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            bool isFinalBlock,
+            Encoding encoding,
+            int keyLengthLimit,
+            int valueLengthLimit,
+            int valueCountLimit)
+        {
+            // We will eventually be comparing against equals and ampersand, so get the encoded
+            // versions of them. For Utf8/ASCII, these are static, so getting the encoding is zero allocation.
+            var equalsEncoded = GetEqualsForEncoding(encoding);
+            var andEncoded = GetAndForEncoding(encoding);
+
+            if (buffer.IsSingleSegment)
+            {
+                TryParseFormValuesFast(buffer.First.Span,
```

Comment:
Trying to make this all static is a mess. Nor does it save you anything when you didn't make the type itself static. If it's all got to be static then make a struct with all the invariant parameters including equalsEncoded, etc..

Reply:
I think that’s fine for now but it’s a waste of an allocation. I agree there are too many arguments here and it’s unclean

Reply:
Yeah I didn't make the final transition to making this all static. There are definitely too many arguments, that's for sure. 

For now, I'll make it a non-static class.